### PR TITLE
Bugprone Narrowing Conversions I

### PR DESCRIPTION
**Pull request by devbali**
Solves bugprone narrowing conversion warning from clang-tidy in many different files.

### CODE REVIEW

File: src/DebugCmds.cc [hunk 1]
```diff
@@ -667,10 +667,10 @@ int dbg_cmd_list(DebugCmd cmd, const vector<string>& args)
 		pre_offset = 0;
 		}
 
-	if ( (int) pre_offset +
-	     (int) g_debugger_state.last_loc.first_line -
-	     (int) CENTER_IDX < 0 )
-		pre_offset = CENTER_IDX - g_debugger_state.last_loc.first_line;
+	if ( static_cast<int> (pre_offset) +
```
Comment (timwoj) [inline src/DebugCmds.cc]: Style nit: remove the space between `>` and `(` here and below.

File: src/Dict.cc [hunk 2]
```diff
@@ -625,7 +625,7 @@ unsigned int Dictionary::MemoryAllocation() const
 		size += pad_size(num_buckets2 * sizeof(PList<DictEntry>*));
 		}
 
-	return size;
+	return static_cast<unsigned int> (size);
```
Comment (timwoj) [inline src/Dict.cc]: Same style nit about the space

File: src/Frame.cc [hunk 3]
```diff
@@ -528,7 +528,7 @@ Frame::UnserializeIDList(const broker::vector& data)
 			}
 
 		ID* id = new ID(has_name->c_str(), SCOPE_FUNCTION, false);
-		id->SetOffset(*has_offset);
+		id->SetOffset( static_cast<int> (*has_offset));
```
Comment (timwoj) [inline src/Frame.cc]: Remove extra spaces here.

File: src/List.h [hunk 4]
```diff
@@ -123,7 +123,7 @@ class List {
 		}
 
 	// Return nth ent of list (do not remove).
-	T& operator[](int i) const
+	T& operator[](unsigned int i) const
```
Comment (timwoj) [inline src/List.h]: I'd be fine with changing this one to a `size_t`. That's how the standard implements it.

**Comment by 0xxon** [comment]
To ask the obvious question - last time we did this, we reverted the change afterwards, due to external breakage. See https://github.com/zeek/zeek/issues/676.

Since this would be the same - and once again probably break just about anything that uses our APIs - I think we should first decide if we actually want to make this kind of change (and if yes, perhaps also re-revert the changes made in https://github.com/zeek/zeek/issues/676).

**Comment by 0xxon** [comment]
And yes, the changes required to external modules are not so huge - but it was the same last time. And modules would probably have to have logic to support both API versions for a while, which increases complexity.

**Comment by jsiwek** [comment]
I don't think there's a blanket statement about whether to do such changes.  It depends case-by-case (I didn't yet look closely at each one here) and it can be a nuanced decision tree:

* Where in the codebase is the change?  Reflects how likely external plugins will be effected.  A change in the DNS parser: probably do whatever.  A static function used by the script interpreter debugger: double-whatever.  A change in some general utility class/function like `ODesc::AddBytes`, or `Val` types like `BroString::Render`, maybe be more careful and move on to the next questions.

* What kind of change is being made?  A change from `int` to `size_t` likely won't break as hard as changing from `int*` to `size_t*` (or references).  The former likely just kicks the implicit integral conversion and narrowing conversion issue into the external plugin's code instead of in ours.  The later change is going to break builds though (this was the case in the previously reverted change IIRC).  In either case, I'd favor just doing a deprecation period to help avoid problems, but more so with the later (if for some reason there's other complications that make the deprecation process burdensome for us to maintain).

* How risky is the warning in the first place?  I don't think people are ever going to pass enough bytes into `ODesc::AddBytes` for the narrowing of some type (probably `size_t` or `int`) into `unsigned int` to cause a problem.

High risk = change warranted

Low risk = if performing the change is not likely to cause problems for others, then sure, change it, otherwise I'd say it's creating extra noise and effort for not much benefit.  e.g. in the `ODesc::AddBytes` case, I'd probably wait to see someone present a valid use-case where they do need to work with such a large number of bytes for the change in type to matter.

Changing commonly-traveled API = maybe worth the caution/vigilance of keeping the old function/signature around in deprecated state.  If it's difficult to do that for some reason and there's a lot of risk associated with the bug-type/warning, then maybe fine to just change outright with no deprecation period.

Hope that logic helps/resonates.

**Comment by rsmmr** [comment]
It sounds like we need a broader discussion here on (1) what our overall target is for clang-tidy checks, and (2) and how to generally approach such changes in terms of API stability. I'll put that on the list for after the holidays.

**Comment by rsmmr** [comment]
Conclusion is that we prefer to be more conservative in terms of what we fix: Let's inspect the cases reported and focus on fixing (1) critical problems and (2) obvious stuff without side effects; and then otherwise resort to marking things as  `// NOLINT` so that clang-tidy will no longer report them. Then we can start monitoring new code more closely. 

Closing this PR for now.

**Pull request closed without merging**